src/sync.rs

```diff
@@ -0,0 +1,302 @@
+                if *count == 0 {
+                    entry.remove();
+                }
+                true
+            }
+            Entry::Vacant(_) => false,
+        }
+    }
+
+    #[inline]
+    /// Put the reader in the queue waiting for the lock.
+    fn rwlock_enqueue_reader(&mut self, id: RwLockId, reader: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(this.rwlock_is_write_locked(id), "queueing on not write locked lock");
+        this.machine.threads.sync.rwlocks[id].reader_queue.push_back(reader);
+    }
+
+    #[inline]
+    /// Take the reader out the queue waiting for the lock.
+    fn rwlock_dequeue_reader(&mut self, id: RwLockId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].reader_queue.pop_front()
+    }
+
+    #[inline]
+    /// Lock by setting the writer that owns the lock.
+    fn rwlock_writer_set(&mut self, id: RwLockId, writer: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(!this.rwlock_is_locked(id), "the lock is already locked");
+        this.machine.threads.sync.rwlocks[id].writer = Some(writer);
+    }
+
+    #[inline]
+    /// Try removing the writer.
+    fn rwlock_writer_remove(&mut self, id: RwLockId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.take()
+    }
+
+    #[inline]
+    /// Put the writer in the queue waiting for the lock.
+    fn rwlock_enqueue_writer(&mut self, id: RwLockId, writer: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(this.rwlock_is_locked(id), "queueing on unlocked lock");
+        this.machine.threads.sync.rwlocks[id].writer_queue.push_back(writer);
+    }
+
+    #[inline]
+    /// Take the writer out the queue waiting for the lock.
+    fn rwlock_dequeue_writer(&mut self, id: RwLockId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer_queue.pop_front()
+    }
+
+    #[inline]
+    /// Create state for a new conditional variable.
+    fn condvar_create(&mut self) -> CondvarId {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.condvars.push(Default::default())
+    }
+
+    #[inline]
+    /// Is the conditional variable awaited?
+    fn condvar_is_awaited(&mut self, id: CondvarId) -> bool {
+        let this = self.eval_context_mut();
+        !this.machine.threads.sync.condvars[id].waiters.is_empty()
+    }
+
+    /// Mark that the thread is waiting on the conditional variable.
+    fn condvar_wait(&mut self, id: CondvarId, thread: ThreadId, mutex: MutexId) {
+        let this = self.eval_context_mut();
+        let waiters = &mut this.machine.threads.sync.condvars[id].waiters;
+        assert!(waiters.iter().all(|waiter| waiter.thread != thread), "thread is already waiting");
+        waiters.push_back(CondvarWaiter { thread, mutex, timeout: None });
```

Comment:
Isn't this the only place where a `CondvarWaiter` is created? So, is `timeout` always `None`?